include/teb_local_planner/g2o_types/edge_velocity.h

```diff
@@ -251,9 +251,20 @@ class EdgeVelocityHolonomic : public BaseTebMultiEdge<3, double>
     double vx = r_dx / deltaT->estimate();
     double vy = r_dy / deltaT->estimate();
     double omega = g2o::normalize_theta(conf2->theta() - conf1->theta()) / deltaT->estimate();
-    
+
+    double max_vel_trans = std::max(std::max(cfg_->robot.max_vel_x, cfg_->robot.max_vel_y), cfg_->robot.max_vel_trans);
+
+    double max_vel_trans_remaining;
+    if (cfg_->optim.norm_vel_trans == 1) {
+      max_vel_trans_remaining = max_vel_trans - std::abs(vx);  // L1 norm
+    }
+    else {
+      max_vel_trans_remaining = std::sqrt(max_vel_trans * max_vel_trans - vx * vx); // L2 norm
+    }
+    double max_vel_y = std::min(max_vel_trans_remaining, cfg_->robot.max_vel_y);
+
     _error[0] = penaltyBoundToInterval(vx, -cfg_->robot.max_vel_x_backwards, cfg_->robot.max_vel_x, cfg_->optim.penalty_epsilon);
-    _error[1] = penaltyBoundToInterval(vy, cfg_->robot.max_vel_y, 0.0); // we do not apply the penalty epsilon here, since the velocity could be close to zero
+    _error[1] = penaltyBoundToInterval(vy, max_vel_y, 0.0); // we do not apply the penalty epsilon here, since the velocity could be close to zero
     _error[2] = penaltyBoundToInterval(omega, cfg_->robot.max_vel_theta,cfg_->optim.penalty_epsilon);
```

Comment:
S / Q:  cannot just add a constraint here like hypot(vx, vy) <= max_vel_trans?  I think it should work, and will be way simpler  :thinking:

Reply:
Do you mean to add a new `_error[3]`?
For that, we need to change the optimization, by changing the matrix information size [here](https://github.com/rst-tu-dortmund/teb_local_planner/blob/melodic-devel/src/optimal_planner.cpp#L757), the `EdgeVelocityHolonomic` template to accommodate a new variable,  and by adding a new variable to optimize (`vel_trans`, which is the coupled velocities). I have tried that, but the behavior is very similar to the current L2 implementation.

Reply:
but,,,  isn't much cleaner than the added implementation?  what is weird is that your constraint seems to subordinate vy to vx,,, what is not the intended behavior
adding _error[3] would be impartial

Reply:
mmm...  better;  
btw, is L1 and L2 behavior anyhow different?  cause as I see this, there's only one correct behavior (what I mention before with the additional constraint)

Reply:
the behavior is the same for both L1 and L2.
I will remove L1 since even the computation is the same now and L2 will always try to optimize to the actual max_vel_trans (L2)